apm-agent-plugins/apm-profiling-plugin/src/main/java/co/elastic/apm/agent/profiler/NativeThreadIdToJavaThreadMapper.java

```diff
@@ -0,0 +1,61 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 - 2019 Elastic and contributors
+ * %%
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.profiler;
+
+import co.elastic.apm.agent.profiler.asyncprofiler.AsyncProfiler;
+import co.elastic.apm.agent.profiler.collections.LongHashSet;
+import com.blogspot.mydailyjava.weaklockfree.DetachedThreadLocal;
+import com.blogspot.mydailyjava.weaklockfree.WeakConcurrentMap;
+
+import java.util.Map;
+
+public class NativeThreadIdToJavaThreadMapper {
+
+    private final DetachedThreadLocal<Long> threadToNativeThread = new DetachedThreadLocal<>(DetachedThreadLocal.Cleaner.INLINE);
+
+    /**
+     * Returns the native thread id of the current thread
+     */
+    public long getNativeThreadId() {
+        Long nativeThreadId = threadToNativeThread.get();
+        if (nativeThreadId == null) {
+            nativeThreadId = AsyncProfiler.getInstance().getNativeThreadId();
+            threadToNativeThread.set(nativeThreadId);
+        }
+        return nativeThreadId;
+    }
```

Comment:
Matching Java threads by native id looks a bit hacky. It probably makes sense for async-profiler to record Java thread id, too.